domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+	// can contain any unicode rune from the letter class and may contain zero
+	// or more of .,+ or - runes as long as they are don't appear at the start
+	// or end of the username.
+	usernameValidationRegex = "^([\\pL\\pN]|[\\pL\\pN][\\pL\\pN.+-]{0,253}[\\pL\\pN])$"
+)
+
+// NewService returns a new Service for interacting with the underlying state.
+func NewService(st State) *Service {
+	return &Service{st: st}
+}
+
+// GetUser will find and return the user associated with name. If there is no
+// user for the user name then an error that satisfies usererrors.NotFound will
+// be returned. If supplied with a invalid user name then a error that satisfies
+// usererrors.
+//
+// GetUser will not return users that have been previously deleted or removed.
+func (s *Service) GetUser(
+	ctx context.Context,
+	name string,
+) (user.User, error) {
+	if err := ValidateUsername(name); err != nil {
+		return user.User{}, fmt.Errorf("username %q: %w", name, err)
+	}
+
+	u, err := s.st.GetUser(ctx, name)
+	if err != nil {
+		return user.User{}, fmt.Errorf("getting user %q: %w", name, err)
+	}
+
+	return u, nil
+}
+
+// ValidateUsername takes a user name and validates that the user name is
+// conformant to our regex rules defined in usernameValidationRegex. If a user
+// name is not valid a error is returned that satisfies
+// usererrors.UsernameNotValid.
+//
+// Usernames must be 1 or more runes long, can contain any unicode rune from the
+// letter class and may contain zero or more of .,+ or - runes as long as they
+// are don't appear at the start or end of the username.
+func ValidateUsername(name string) error {
+	regex, err := regexp.Compile(usernameValidationRegex)
+	if err != nil {
+		return fmt.Errorf("compiling user name validation regex %q: %w",
+			usernameValidationRegex, err,
+		)
+	}
+	if !regex.MatchString(name) {
+		return fmt.Errorf("%w %q", usererrors.UsernameNotValid, name)
+	}
+	return nil
+}
+
+// AddUser will add a new user to the database. If the user already exists an
+// error that satisfies usererrors.AlreadyExists will be returned.
+func (s *Service) AddUser(ctx context.Context, name, displayName, password, creator string) error {
+	err := ValidateUsername(name)
+	if err != nil {
+		return fmt.Errorf("username %q: %w", name, err)
+	}
+
+	err = s.st.AddUser(ctx, name, displayName, password, creator)
+	if err != nil {
+		return fmt.Errorf("adding user %q: %w", name, err)
+	}
+	return nil
+}
+
+// RemoveUser marks the user as deleted. This obviates the ability of a user
+// to function, but keeps the userDoc retaining provenance, i.e. auditing.
+func (s *Service) RemoveUser(ctx context.Context, name string) error {
+	if err := ValidateUsername(name); err != nil {
+		return fmt.Errorf("username %q: %w", name, err)
+	}
+	err := s.st.RemoveUser(ctx, name)
+	if err != nil {
+		return fmt.Errorf("removing user %q: %w", name, err)
+	}
+	return nil
+}
+
+// SetPassword removes any active activation keys and sets the user password
```

Comment:
We need to document the error cases here. What happens if the username is not valid or doesn't exist?